File: ports/python/test_complex.py
from dataclasses import dataclass, field
from typing import List, Dict, Any, Tuple, Callable
import std as benc

@dataclass
class SubSubItem:
    sub_sub_id: str = ""
    sub_sub_data: bytes = b""
    
    def size_plain(self) -> int:
        return benc.size_string(self.sub_sub_id) + benc.size_bytes(self.sub_sub_data)
    
    def marshal_plain(self, n: int, b: bytearray) -> int:
        n = benc.marshal_string(n, b, self.sub_sub_id)
        return benc.marshal_bytes(n, b, self.sub_sub_data)
    
    def unmarshal_plain(self, n: int, b: bytes) -> int:
        n, self.sub_sub_id = benc.unmarshal_string(n, b)
        n, self.sub_sub_data = benc.unmarshal_bytes_cropped(n, b)
        return n

@dataclass
class SubItem:
    sub_id: int = 0
    description: str = ""
    sub_items: List[SubSubItem] = field(default_factory=list)
    
    def size_plain(self) -> int:
        return (
            benc.size_int32() + 
            benc.size_string(self.description) + 
            benc.size_slice(self.sub_items, lambda x: x.size_plain())
        )
    
    def marshal_plain(self, n: int, b: bytearray) -> int:
        n = benc.marshal_int32(n, b, self.sub_id)
        n = benc.marshal_string(n, b, self.description)
        return benc.marshal_slice(n, b, self.sub_items, lambda n, b, item: item.marshal_plain(n, b))
    
    def unmarshal_plain(self, n: int, b: bytes) -> int:
        n, self.sub_id = benc.unmarshal_int32(n, b)
        n, self.description = benc.unmarshal_string(n, b)
        
        def unmarshal_subsubitem(n: int, b: bytes) -> Tuple[int, SubSubItem]:
            item = SubSubItem()
            n = item.unmarshal_plain(n, b)
            return n, item
        
        n, self.sub_items = benc.unmarshal_slice(n, b, unmarshal_subsubitem)
        return n

@dataclass
class SubComplexData:
    sub_id: int = 0
    sub_title: str = ""
    sub_binary_data: List[bytes] = field(default_factory=list)
    sub_items: List[SubItem] = field(default_factory=list)
    sub_metadata: Dict[str, str] = field(default_factory=dict)
    
    def size_plain(self) -> int:
        return (
            benc.size_int32() +
            benc.size_string(self.sub_title) +
            benc.size_slice(self.sub_binary_data, benc.size_bytes) +
            benc.size_slice(self.sub_items, lambda x: x.size_plain()) +
            benc.size_map(self.sub_metadata, benc.size_string, benc.size_string)
        )
    
    def marshal_plain(self, n: int, b: bytearray) -> int:
        n = benc.marshal_int32(n, b, self.sub_id)
        n = benc.marshal_string(n, b, self.sub_title)
        n = benc.marshal_slice(n, b, self.sub_binary_data, benc.marshal_bytes)
        n = benc.marshal_slice(n, b, self.sub_items, lambda n, b, item: item.marshal_plain(n, b))
        return benc.marshal_map(n, b, self.sub_metadata, benc.marshal_string, benc.marshal_string)
    
    def unmarshal_plain(self, n: int, b: bytes) -> int:
        n, self.sub_id = benc.unmarshal_int32(n, b)
        n, self.sub_title = benc.unmarshal_string(n, b)
        n, self.sub_binary_data = benc.unmarshal_slice(n, b, benc.unmarshal_bytes_cropped)
        
        def unmarshal_subitem(n: int, b: bytes) -> Tuple[int, SubItem]:
            item = SubItem()
            n = item.unmarshal_plain(n, b)
            return n, item
        
        n, self.sub_items = benc.unmarshal_slice(n, b, unmarshal_subitem)
        n, self.sub_metadata = benc.unmarshal_map(n, b, benc.unmarshal_string, benc.unmarshal_string)
        return n

@dataclass
class ComplexData:
    id: int = 0
    title: str = ""
    items: List[SubItem] = field(default_factory=list)
    metadata: Dict[str, int] = field(default_factory=dict)
    sub_data: SubComplexData = field(default_factory=SubComplexData)
    large_binary_data: List[bytes] = field(default_factory=list)
    huge_list: List[int] = field(default_factory=list)
    
    def size_plain(self) -> int:
        return (
            benc.size_int64() +
            benc.size_string(self.title) +
            benc.size_slice(self.items, lambda x: x.size_plain()) +
            benc.size_map(self.metadata, benc.size_string, benc.size_int32) +
            self.sub_data.size_plain() +
            benc.size_slice(self.large_binary_data, benc.size_bytes) +
            benc.size_fixed_slice(self.huge_list, benc.size_int64())
        )
    
    def marshal_plain(self, n: int, b: bytearray) -> int:
        n = benc.marshal_int64(n, b, self.id)
        n = benc.marshal_string(n, b, self.title)
        n = benc.marshal_slice(n, b, self.items, lambda n, b, item: item.marshal_plain(n, b))
        n = benc.marshal_map(n, b, self.metadata, benc.marshal_string, benc.marshal_int32)
        n = self.sub_data.marshal_plain(n, b)
        n = benc.marshal_slice(n, b, self.large_binary_data, benc.marshal_bytes)
        return benc.marshal_slice(n, b, self.huge_list, benc.marshal_int64)
    
    def unmarshal_plain(self, n: int, b: bytes) -> int:
        n, self.id = benc.unmarshal_int64(n, b)
        n, self.title = benc.unmarshal_string(n, b)
        
        def unmarshal_subitem(n: int, b: bytes) -> Tuple[int, SubItem]:
            item = SubItem()
            n = item.unmarshal_plain(n, b)
            return n, item
        
        n, self.items = benc.unmarshal_slice(n, b, unmarshal_subitem)
        n, self.metadata = benc.unmarshal_map(n, b, benc.unmarshal_string, benc.unmarshal_int32)
        n = self.sub_data.unmarshal_plain(n, b)
        n, self.large_binary_data = benc.unmarshal_slice(n, b, benc.unmarshal_bytes_cropped)
        n, self.huge_list = benc.unmarshal_slice(n, b, benc.unmarshal_int64)
        return n

def create_test_data() -> ComplexData:
    """Create test data with nested structures"""
    return ComplexData(
        id=12345,
        title="Example Complex Data",
        items=[
            SubItem(
                sub_id=1,
                description="SubItem 1",
                sub_items=[
                    SubSubItem(
                        sub_sub_id="subsub1",
                        sub_sub_data=bytes([0x01, 0x02, 0x03])
                    )
                ]
            )
        ],
        metadata={
            "key1": 10,
            "key2": 20
        },
        sub_data=SubComplexData(
            sub_id=999,
            sub_title="Sub Complex Data",
            sub_binary_data=[
                bytes([0x11, 0x22, 0x33]),
                bytes([0x44, 0x55, 0x66])
            ],
            sub_items=[
                SubItem(
                    sub_id=2,
                    description="SubItem 2",
                    sub_items=[
                        SubSubItem(
                            sub_sub_id="subsub2",
                            sub_sub_data=bytes([0xAA, 0xBB, 0xCC])
                        )
                    ]
                )
            ],
            sub_metadata={
                "meta1": "value1",
                "meta2": "value2"
            }
        ),
        large_binary_data=[
            bytes([0xFF, 0xEE, 0xDD])
        ],
        huge_list=[1000000, 2000000, 3000000]
    )

def test_complex_data() -> None:
    """Test complex data structure serialization/deserialization"""
    # Create test data
    original_data = create_test_data()
    
    # Calculate size and create buffer
    size = original_data.size_plain()
    buffer = bytearray(size)
    
    # Marshal the data
    marshal_result = original_data.marshal_plain(0, buffer)
    assert marshal_result == size, f"Marshal failed: expected {size} bytes, got {marshal_result}"
    
    # Unmarshal the data
    unmarshaled_data = ComplexData()
    unmarshal_result = unmarshaled_data.unmarshal_plain(0, buffer)
    assert unmarshal_result == size, f"Unmarshal failed: expected {size} bytes, got {unmarshal_result}"
    
    # Verify data integrity
    assert original_data.id == unmarshaled_data.id
    assert original_data.title == unmarshaled_data.title
    assert original_data.metadata == unmarshaled_data.metadata
    assert original_data.sub_data.sub_id == unmarshaled_data.sub_data.sub_id
    assert original_data.sub_data.sub_title == unmarshaled_data.sub_data.sub_title
    assert original_data.sub_data.sub_metadata == unmarshaled_data.sub_data.sub_metadata
    assert original_data.huge_list == unmarshaled_data.huge_list
    
    # Verify nested structures
    assert len(original_data.items) == len(unmarshaled_data.items)
    assert len(original_data.sub_data.sub_items) == len(unmarshaled_data.sub_data.sub_items)
    assert len(original_data.large_binary_data) == len(unmarshaled_data.large_binary_data)
    
    # Verify binary data
    for i, (orig, unmarshaled) in enumerate(zip(original_data.large_binary_data, unmarshaled_data.large_binary_data)):
        assert orig == unmarshaled, f"Binary data mismatch at index {i}"
    
    print("✓ Complex data test passed!")

if __name__ == "__main__":
    test_complex_data()

File: ports/rust/tests/test_complex.rs
#[cfg(test)]
mod complex_tests {
    use benc::*;
    use std::collections::HashMap;

    #[derive(Debug, PartialEq, Clone)]
    struct ComplexData {
        id: i64,
        title: String,
        items: Vec<SubItem>,
        metadata: HashMap<String, i32>,
        sub_data: SubComplexData,
        large_binary_data: Vec<Vec<u8>>,
        huge_list: Vec<i64>,
    }

    impl ComplexData {
        fn size(&self) -> usize {
            size_int(self.id) +
            size_string(&self.title) +
            size_slice(&self.items, |item| item.size()) +
            size_map(&self.metadata, |k| size_string(k), |v| size_int(*v as i64)) + // Convert i32 to i64 for size calculation
            self.sub_data.size() +
            size_slice(&self.large_binary_data, |b| size_bytes(b)) +
            size_slice(&self.huge_list, |v| size_int(*v)) // Use variable-length encoding for consistency
        }

        fn marshal(&self, writer: &mut &mut [u8]) {
            marshal_int(self.id, writer);
            marshal_string(&self.title, writer);
            marshal_slice(&self.items, writer, |item, w| item.marshal(w));
            marshal_map(&self.metadata, writer, 
                |k, w| marshal_string(k, w), 
                |v, w| marshal_int(*v as i64, w) // Convert i32 to i64 for marshaling
            );
            self.sub_data.marshal(writer);
            marshal_slice(&self.large_binary_data, writer, |b, w| marshal_bytes(b, w));
            marshal_slice(&self.huge_list, writer, |v, w| marshal_int(*v, w));
        }

        fn unmarshal(reader: &mut &[u8]) -> Result<Self> {
            Ok(Self {
                id: unmarshal_int(reader)?,
                title: unmarshal_string(reader)?.to_string(),
                items: unmarshal_slice(reader, SubItem::unmarshal)?,
                metadata: unmarshal_map(
                    reader,
                    |r| unmarshal_string(r).map(|s| s.to_string()),
                    |r| unmarshal_int(r).map(|v| v as i32) // Convert back to i32 after unmarshaling
                )?,
                sub_data: SubComplexData::unmarshal(reader)?,
                large_binary_data: unmarshal_slice(reader, |r| 
                    unmarshal_bytes_copied(r).map(|v| v)
                )?,
                huge_list: unmarshal_slice(reader, unmarshal_int)?,
            })
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    struct SubItem {
        sub_id: i32,
        description: String,
        sub_items: Vec<SubSubItem>,
    }

    impl SubItem {
        fn size(&self) -> usize {
            size_int(self.sub_id as i64) + // Use variable-length encoding
            size_string(&self.description) +
            size_slice(&self.sub_items, |item| item.size())
        }

        fn marshal(&self, writer: &mut &mut [u8]) {
            marshal_int(self.sub_id as i64, writer); // Use variable-length encoding
            marshal_string(&self.description, writer);
            marshal_slice(&self.sub_items, writer, |item, w| item.marshal(w));
        }

        fn unmarshal(reader: &mut &[u8]) -> Result<Self> {
            Ok(Self {
                sub_id: unmarshal_int(reader)? as i32, // Convert back to i32
                description: unmarshal_string(reader)?.to_string(),
                sub_items: unmarshal_slice(reader, SubSubItem::unmarshal)?,
            })
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    struct SubSubItem {
        sub_sub_id: String,
        sub_sub_data: Vec<u8>,
    }

    impl SubSubItem {
        fn size(&self) -> usize {
            size_string(&self.sub_sub_id) +
            size_bytes(&self.sub_sub_data)
        }

        fn marshal(&self, writer: &mut &mut [u8]) {
            marshal_string(&self.sub_sub_id, writer);
            marshal_bytes(&self.sub_sub_data, writer);
        }

        fn unmarshal(reader: &mut &[u8]) -> Result<Self> {
            Ok(Self {
                sub_sub_id: unmarshal_string(reader)?.to_string(),
                sub_sub_data: unmarshal_bytes_copied(reader)?,
            })
        }
    }

    #[derive(Debug, PartialEq, Clone)]
    struct SubComplexData {
        sub_id: i32,
        sub_title: String,
        sub_binary_data: Vec<Vec<u8>>,
        sub_items: Vec<SubItem>,
        sub_metadata: HashMap<String, String>,
    }

    impl SubComplexData {
        fn size(&self) -> usize {
            size_int(self.sub_id as i64) + // Use variable-length encoding
            size_string(&self.sub_title) +
            size_slice(&self.sub_binary_data, |b| size_bytes(b)) +
            size_slice(&self.sub_items, |item| item.size()) +
            size_map(&self.sub_metadata, |k| size_string(k), |v| size_string(v))
        }

        fn marshal(&self, writer: &mut &mut [u8]) {
            marshal_int(self.sub_id as i64, writer); // Use variable-length encoding
            marshal_string(&self.sub_title, writer);
            marshal_slice(&self.sub_binary_data, writer, |b, w| marshal_bytes(b, w));
            marshal_slice(&self.sub_items, writer, |item, w| item.marshal(w));
            marshal_map(&self.sub_metadata, writer,
                |k, w| marshal_string(k, w),
                |v, w| marshal_string(v, w)
            );
        }

        fn unmarshal(reader: &mut &[u8]) -> Result<Self> {
            Ok(Self {
                sub_id: unmarshal_int(reader)? as i32, // Convert back to i32
                sub_title: unmarshal_string(reader)?.to_string(),
                sub_binary_data: unmarshal_slice(reader, |r| 
                    unmarshal_bytes_copied(r).map(|v| v)
                )?,
                sub_items: unmarshal_slice(reader, SubItem::unmarshal)?,
                sub_metadata: unmarshal_map(
                    reader,
                    |r| unmarshal_string(r).map(|s| s.to_string()),
                    |r| unmarshal_string(r).map(|s| s.to_string())
                )?,
            })
        }
    }

    #[test]
    fn test_complex() {
        let data = ComplexData {
            id: 12345,
            title: "Example Complex Data".to_string(),
            items: vec![SubItem {
                sub_id: 1,
                description: "SubItem 1".to_string(),
                sub_items: vec![SubSubItem {
                    sub_sub_id: "subsub1".to_string(),
                    sub_sub_data: vec![0x01, 0x02, 0x03],
                }],
            }],
            metadata: {
                let mut map = HashMap::new();
                map.insert("key1".to_string(), 10);
                map.insert("key2".to_string(), 20);
                map
            },
            sub_data: SubComplexData {
                sub_id: 999,
                sub_title: "Sub Complex Data".to_string(),
                sub_binary_data: vec![
                    vec![0x11, 0x22, 0x33],
                    vec![0x44, 0x55, 0x66],
                ],
                sub_items: vec![SubItem {
                    sub_id: 2,
                    description: "SubItem 2".to_string(),
                    sub_items: vec![SubSubItem {
                        sub_sub_id: "subsub2".to_string(),
                        sub_sub_data: vec![0xAA, 0xBB, 0xCC],
                    }],
                }],
                sub_metadata: {
                    let mut map = HashMap::new();
                    map.insert("meta1".to_string(), "value1".to_string());
                    map.insert("meta2".to_string(), "value2".to_string());
                    map
                },
            },
            large_binary_data: vec![vec![0xFF, 0xEE, 0xDD]],
            huge_list: vec![1000000, 2000000, 3000000],
        };

        // Calculate size and create buffer
        let size = data.size();
        let mut buf = vec![0u8; size];
        let mut writer = &mut buf[..];
        
        // Marshal the data
        data.marshal(&mut writer);
        assert_eq!(writer.len(), 0, "Buffer not fully written");

        // Unmarshal the data
        let mut reader = &buf[..];
        let ret_data = ComplexData::unmarshal(&mut reader).unwrap();
        
        // Verify all data was read
        assert_eq!(reader.len(), 0, "Buffer not fully read");
        
        // Verify data matches
        assert_eq!(data, ret_data);
    }
}